### PR TITLE
Fix for zfs partition deletion

### PR DESCRIPTION
**Pull request by dalto8**
kpmcore uses zfs to destroy the zpool on a zfs partition.

In the case of Calamares, most ISOs don't include zfs or the needed kernel modules.  This means that when Calamares tries to delete an existing zfs partition, it results in error.

This change checks to see if there is working zfs environment.  If there isn't, it uses `sfdisk` to manually delete the partition.  `sfdisk` was chosen since it is used by kpmcore and shouldn't add any additional dependencies.

### CODE REVIEW

**Comment by kkofler** [comment]
I think this belongs into kpmcore, not Calamares.

**Comment by dalto8** [comment]
> I think this belongs into kpmcore, not Calamares.

It is an interesting question.  Should kpmcore check for zfs presence or should it rely on anyone leveraging the functionality to implement it if they call it?  I am not sure what their position on that is.

Either way, Calamares users will continue to have failed installs in the meantime.

**Comment by kkofler** [comment]
@stikonas: What do you think?

(FYI, @stikonas is the kpmcore maintainer.)

File: src/modules/partition/jobs/DeletePartitionJob.cpp [hunk 1]
```diff
@@ -49,6 +51,29 @@ DeletePartitionJob::prettyStatusMessage() const
 Calamares::JobResult
 DeletePartitionJob::exec()
 {
+    // Special handling for zfs
```
Comment (adriaandegroot) [inline src/modules/partition/jobs/DeletePartitionJob.cpp]: Typically I'd split this off into a separate function -- probably a static free function -- so that the overall flow is easier to read; something along the lines of

```
if (isZfs(m_partition)) { return removeZfs(m_partition); }
```

Then bung the rest of this code into those functions.

File: src/modules/partition/jobs/DeletePartitionJob.cpp [hunk 2]
```diff
@@ -49,6 +51,29 @@ DeletePartitionJob::prettyStatusMessage() const
...
+
+        if ( r.getExitCode() != 0 )
+        {
+            r = CalamaresUtils::System::instance()->runCommand( { "sfdisk",
+                                                                  "--delete",
+                                                                  "--force",
+                                                                  m_partition->devicePath(),
+                                                                  QString::number( m_partition->number() ) },
+                                                                std::chrono::seconds( 5 ) );
+            if ( r.getExitCode() != 0 )
```
Comment (adriaandegroot) [inline src/modules/partition/jobs/DeletePartitionJob.cpp]: Please use `{ }` around every block (Calamares **has** had bugs because of suggestive indentation and missing braces, and it's made worse by mixing Python and C++ where "suggestive indentation" is "meaningful indentation" half the time).

File: src/modules/partition/jobs/DeletePartitionJob.cpp [hunk 3]
```diff
@@ -49,6 +51,29 @@ DeletePartitionJob::prettyStatusMessage() const
...
+                return Calamares::JobResult::error( "message",
```
Comment (adriaandegroot) [inline src/modules/partition/jobs/DeletePartitionJob.cpp]: These are user-visible, so you need to apply `QObject::tr()` to them (if you move this code to a static free function, it becomes a bit wordy, as

```
QCoreApplication::translate( DeletePartitionJob::staticMetaObject.className(), "message" )
```

.. not that I think "message" is a good title for an error message describing the failure to delete a partition.

**Review by adriaandegroot — COMMENTED** [review]
I share the feeling that this belongs in kpmcore -- you ought to be able to delete a partition even if the FS on that partition is unsupported .. um, hang on. This is weird **anyway** because a partition would typically belong to a pool, not to a single ZFS filesystem. OTOH, it'd be nice to have delete-ZFS-partition working even before a new kpmcore release.

There's comments on the code, though, regardless of whether this belongs here or not.

**Comment by stikonas** [comment]
> @stikonas: What do you think?
> 
> (FYI, @stikonas is the kpmcore maintainer.)

Yes, I think it would be good to get this into KPMcore. KPMcore releases are fairly frequent these days. Next release is in December and feature freeze is 12 November.

**Comment by dalto8** [comment]
> > @stikonas: What do you think?
> > (FYI, @stikonas is the kpmcore maintainer.)
> 
> Yes, I think it would be good to get this into KPMcore. 

Probably not the right place to have this discussion, but I don't think kpmcore should need to do anything this complicated.  Since kpmcore can't understand intention, it should alwyas just delete the partition and not worry about the fact that it is a zfs partition.  What it currently does, forcibly destroying a zpool based on the partition label is unbelievably dangerous.  What if someone is trying to delete a partition that is part of a mirror or raid set and not expecting the entire pool to be destroyed?  Worse, what if the partition is one that has been removed from the pool and still has the label of the pool?

I am not that familiar with kpmcore but I guess that would be just removing the `remove()` function from the zfs implementation and letting the base class handle the remove?

**Comment by stikonas** [comment]
> > > @stikonas: What do you think?
> > > (FYI, @stikonas is the kpmcore maintainer.)
> > 
> > 
> > Yes, I think it would be good to get this into KPMcore.
> 
> Probably not the right place to have this discussion, but I don't think kpmcore should need to do anything this complicated. Since kpmcore can't understand intention, it should alwyas just delete the partition and not worry about the fact that it is a zfs partition. What it currently does, forcibly destroying a zpool based on the partition label is unbelievably dangerous. What if someone is trying to delete a partition that is part of a mirror or raid set and not expecting the entire pool to be destroyed? Worse, what if the partition is one that has been removed from the pool and still has the label of the pool?
> 
> I am not that familiar with kpmcore but I guess that would be just removing the `remove()` function from the zfs implementation and letting the base class handle the remove?

I don't remember why exactly there is zpool destruction. But It looks like it is indeed too much. But maybe there was an issue that if zpool if active then `remove()` failed. But I am not really familiar with ZFS, so I might be saying nonsense.

**Comment by adriaandegroot** [comment]
This discussion suggests that what Calamares should do is **ignore** the current KPMCore implementation and always use the `sfdisk remove` approach here (e.g. we can lose the "zpool status" check). That will damage the pool, but avoid whatever thing kpmcore is doing with the pool as a whole.

**Comment by dalto8** [comment]
Let me know if this is more inline with what you were thinking.

**Comment by stikonas** [comment]
And `kpmcore` is now switched to just remove partition without messing up with pools.
